pyteal/ast/subroutine.py

```diff
@@ -911,6 +913,23 @@ class SubroutineEval:
     use_frame_pt: bool = False
     _current_proto: ClassVar[Optional[Proto]] = None
 
+    @staticmethod
+    def _new_abi_instance_from_storage(
+        spec: abi.TypeSpec, storage: FrameVar
+    ) -> abi.BaseType:
+        """
+        This hidden method generates new ABI instance that is tied to the storage: FrameVar as follows:
+        - generates new instance that is based on scratch vars
+        - rewind the new instance to be using storage: FrameVar
+        - rewind the state changed by scratch slot allocation
+        """
+        current_scratch_id = ScratchSlot.nextSlotId
+        with _frame_pointer_context(None):
+            instance = spec.new_instance()
+        instance._stored_value = storage
+        ScratchSlot.reset_slot_numbering(current_scratch_id)
```

Comment:
no, we are fairly sure that this `instance` is for local storage:
- argument of the function
- output result of the function (if it exists)

The purpose here is just to create an instance with storage tied to stack, rather than scratch slot. Thus, we don't need a slot number here, and we reset the slot number after `new_instance` with `None` being the current proto.

Reply:
Ok thx.  It makes sense in the case that frame pointers are allocated - no slots involved at all. But what about the case that we run out of frame pointers and need to switch to slots? It can't be that after the slot allocation for local var's another call is made that would end up errantly re-using these loca slots while their still alive for the calling subroutine?

Reply:
So I think I'm starting to understand that it's safe to reset the numbering when this is called. Any slots being allocated here are or only for decoding of the abi-variable and are immediately discardable after being loaded. So later on, say in the logic of same subroutine, there is no harm to reuse the same slot. Is my understnding correct?

Reply:
>  Any slots being allocated here are or only for decoding of the abi-variable and are immediately discardable after being loaded. 

exactly. we don't care about the scratch slots we allocating here, they are not involved in teal runtime, so your understanding is correct.

Reply:
If it isn't too onerous, I think it would be good to create a unit test, either of `var_n_loaded_scratch()` and/or `evaluate()` (I see several unit tests already exist for this) and/or `_new_abi_instance_from_storage()`. Ideally, this would also provide a test case that we know barfs without this fix, but is shown to work fine, because of the fix.

Do you think that this is worthwhile?

Reply:
sry for late reply, and yes, that makes sense to me, working on a set of unit test to make sure no side effect introduced.

Reply:
yes, I made a testcase deriving from the reproduction in [8f5f61e](https://github.com/algorand/pyteal/pull/684/commits/8f5f61ee796cccf957ab140e899e9b09ed3f726f), confirming local vars should be rightly allocated.